uni-stark/src/prover.rs

```diff
@@ -3,82 +3,74 @@ use alloc::vec::Vec;
 use tracing::{info_span, instrument};
 
 use crate::symbolic_builder::{get_log_quotient_degree, SymbolicAirBuilder};
 use crate::{
-    decompose_and_flatten, Commitments, OpenedValues, PackedChallenge, PackedVal, Proof,
-    ProverConstraintFolder, StarkGenericConfig, ZerofierOnCoset,
+    Commitments, Domain, OpenedValues, PackedChallenge, PackedVal, Proof, ProverConstraintFolder,
+    StarkGenericConfig, Val,
 };
 
 #[instrument(skip_all)]
 pub fn prove<
     SC,
-    #[cfg(debug_assertions)] A: for<'a> Air<crate::check_constraints::DebugConstraintBuilder<'a, SC::Val>>,
+    #[cfg(debug_assertions)] A: for<'a> Air<crate::check_constraints::DebugConstraintBuilder<'a, Val<SC>>>,
     #[cfg(not(debug_assertions))] A,
 >(
     config: &SC,
     air: &A,
     challenger: &mut SC::Challenger,
-    trace: RowMajorMatrix<SC::Val>,
+    trace: RowMajorMatrix<Val<SC>>,
 ) -> Proof<SC>
 where
     SC: StarkGenericConfig,
-    A: Air<SymbolicAirBuilder<SC::Val>> + for<'a> Air<ProverConstraintFolder<'a, SC>>,
+    A: Air<SymbolicAirBuilder<Val<SC>>> + for<'a> Air<ProverConstraintFolder<'a, SC>>,
 {
     #[cfg(debug_assertions)]
     crate::check_constraints::check_constraints(air, &trace);
 
     let degree = trace.height();
     let log_degree = log2_strict_usize(degree);
 
-    let log_quotient_degree = get_log_quotient_degree::<SC::Val, A>(air);
-
-    let g_subgroup = SC::Val::two_adic_generator(log_degree);
+    let log_quotient_degree = get_log_quotient_degree::<Val<SC>, A>(air);
 
     let pcs = config.pcs();
+    let trace_domain = pcs.natural_domain_for_degree(degree);
+
     let (trace_commit, trace_data) =
-        info_span!("commit to trace data").in_scope(|| pcs.commit_batch(trace));
+        info_span!("commit to trace data").in_scope(|| pcs.commit(vec![(trace_domain, trace)]));
 
     challenger.observe(trace_commit.clone());
     let alpha: SC::Challenge = challenger.sample_ext_element();
 
-    let mut trace_ldes = pcs.get_ldes(&trace_data);
-    assert_eq!(trace_ldes.len(), 1);
-    let trace_lde = trace_ldes.pop().unwrap();
+    let quotient_domain =
+        trace_domain.create_disjoint_domain(1 << (log_degree + log_quotient_degree));
 
-    let log_stride_for_quotient = pcs.log_blowup() - log_quotient_degree;
-    let trace_lde_for_quotient = trace_lde.vertically_strided(1 << log_stride_for_quotient, 0);
+    let trace_on_quotient_domain = pcs.get_evaluations_on_domain(&trace_data, 0, quotient_domain);
 
     let quotient_values = quotient_values(
-        config,
         air,
-        log_degree,
-        log_quotient_degree,
-        trace_lde_for_quotient,
+        trace_domain,
+        quotient_domain,
+        trace_on_quotient_domain,
         alpha,
     );
-    let quotient_chunks_flattened = decompose_and_flatten(
-        quotient_values,
-        SC::Challenge::from_base(pcs.coset_shift()),
-        log_quotient_degree,
-    );
+    let quotient_flat = RowMajorMatrix::new_col(quotient_values).flatten_to_base();
```

Comment:
Now flattening and then decomposing instead of the other way around. Idk, could go either way, but this way the decompose is easier to pack